sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/url/UrlExpressionEvalUtils.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.expressions.url
+
+import java.net.{URI, URISyntaxException}
+import java.util.regex.Pattern
+
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.unsafe.types.UTF8String
+
+case class ParseUrlEvaluator(
+    cachedUrl: () => URI,
+    cachedExtractPartFunc: URI => String,
+    cachedPattern: () => Pattern,
```

Comment:
it's better to not put lambda functions here as `ParseUrlEvaluator` will be serialized and sent to the executor side.

Can we pass the necessary basic information so that we can generate these lambda functions within `ParseUrlEvaluator`?

Reply:
Let me try.

Reply:
Updated!
(PS: This is really a `good` suggestion. For myself, the code looks much better now, thanks! ❤️)